### PR TITLE
Adding SDK support for Python 3.8 and 3.9

### PR DESCRIPTION
**Pull request by krithika369**
The Turing Python SDK currently only supports Python version `3.7`. This has been enforced because of cross-version compatibility issues with pickling.

### Background
The Pyfunc ensemblers created using the SDK are saved as pickled artifacts and loaded into the Python server, when deployed as a batch job or a real-time webservice. In some cases (such as, when the Python objects are defined in the main scope; see https://github.com/cloudpipe/cloudpickle/issues/404), the artifacts are unable to be loaded into a different Python version than the version they were created in. For example, an ensembler created using Python 3.8 / 3.9 cannot be served using a webserver built with Python 3.7. To workaround this limitation, both the SDK and the Pyfunc server base images were pinned at Python 3.7. This PR adds support for newer Python versions.

### Solution
When the ensembler is submitted using the SDK, the user's minor python version is captured and saved to the `pyfunc_ensemblers` table. Base images for the batch ensembling jobs and real-time ensemblers will now be created for each supported Python version. At the time of ensembler deployment, the matching base image is chosen.

Alternatively, the base image could have been loaded with different Python versions and the right one activated at the time of ensembler image building and serving. But this would make the base image very bulky (adds another GB to the current size of `1.6` GB).

#### API
* `api/api/specs/ensemblers.yaml` - Add `python_version` as an attribute of the `PyFuncEnsembler`
* `api/db-migrations/000011_add_pyfunc_ensemblers_python_version.up.sql` - Add `python_version` to the DB table, defaulting it to `3.7.*` for existing records.
* `api/turing/config/config.go` - `BaseImageRef` is changed from a string to a map, to hold the image name per minor Python version
* `api/turing/batch/ensembling/runner.go`, `api/turing/service/router_deployment_service.go` - Set the Python version as the base image tag when calling the image builder
* `api/turing/imagebuilder/imagebuilder.go` - Pick the base image using the supplied tag

#### Engines
* `engines/**/Dockerfile` - Use Python-version specific conda environment file
* `engines/**/Makefile` - Add PYTHON_VERSION variable
* `engines/**/env-3.*.yaml` - Split the existing `environment.yaml` into one per supported version
* `engines/**/requirements.txt` - Upgraded some of the requirements to versions supported by all 3 supported Python versions, upgraded `cloudpickle` to `2.0.0`. Pinning `protobuf` to `3.20.1` so the `mlflow` library can work on Python >=3.8.

#### SDK
* `sdk/setup.py` - Update Python versions required
* `sdk/turing/ensembler.py` - During create / update of Pyfunc ensembler, capture the user's Python version in the API call. The version is also added to the `conda_env` before logging with `mlflow`.

#### Other Changes
* Update Github CI workflows to test the SDK / Pyfunc servers and publish the Pyfunc servers for each supported Python minor version
* Update default Helm values to match the new Imagebuilder configs

CC: Merlin Dev (@ariefrahmansyah @tiopramayudi @pradithya)

### CODE REVIEW

File: api/turing/api/appcontext.go [hunk 1]
```diff
@@ -128,6 +131,7 @@ func NewAppContext(
 		batchEnsemblingJobRunner := batchensembling.NewBatchEnsemblingJobRunner(
 			batchEnsemblingController,
 			ensemblingJobService,
+			ensemblersService,
```
Comment (krithika369) [inline api/turing/api/appcontext.go:134]: The batch job runner now needs the EnsemblersService, to query the python version information at the time of image building. See: `api/turing/batch/ensembling/runner.go`

File: engines/pyfunc-ensembler-job/requirements.txt [hunk 2]
```diff
@@ -1,8 +1,9 @@
+cloudpickle==2.0.0
 google-cloud-storage>=1.37.0
 jinjasql==0.1.8
-jinja2==3.0.3  # temporary fix to downgrade to Python 3.7
+jinja2==3.0.3
 mlflow==1.14.1
-numpy==1.17.3
```
Comment (krithika369) [inline engines/pyfunc-ensembler-job/requirements.txt:5]: `numpy 1.17.3` is not available for Python 3.9 (See: https://anaconda.org/anaconda/numpy/files?version=1.17.3)

File: engines/pyfunc-ensembler-service/requirements.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 argparse>=1.4.0
-cloudpickle==1.2.2
-orjson==2.6.8
+cloudpickle==2.0.0
+orjson==3.6.8
```
Comment (krithika369) [inline engines/pyfunc-ensembler-service/requirements.txt:3]: orjson `2.6.8` is not available for Python 3.9

File: infra/charts/turing/values.yaml [hunk 4]
```diff
@@ -125,7 +125,8 @@ turing:
         BuildNamespace: default
         BuildTimeoutDuration: 20m
         DestinationRegistry: ghcr.io
-        BaseImageRef: ghcr.io/gojek/turing/pyfunc-ensembler-service:latest
+        BaseImageRef:
+          3.7.*: ghcr.io/gojek/turing/pyfunc-ensembler-service:latest
```
Comment (krithika369) [inline infra/charts/turing/values.yaml:129]: Updating the base values to match the new format. Will capture the mapping for the new python versions and the corresponding public artifacts in chart version `0.2.9`.

File: sdk/turing/ensembler.py [hunk 5]
```diff
@@ -197,6 +203,14 @@ def artifact_uri(self) -> str:
     @artifact_uri.setter
     def artifact_uri(self, artifact_uri):
         self._artifact_uri = artifact_uri
+    
+    @property
+    def python_version(self) -> str:
+        return self._python_version
+
+    @python_version.setter
+    def python_version(self, python_version):
```
Comment (terryyylim) [inline sdk/turing/ensembler.py:212]: nit: `def python_version(self, python_version: str) -> None:`

File: sdk/turing/ensembler.py [hunk 6]
```diff
@@ -341,9 +358,33 @@ def create(
...
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+    elif not isinstance(conda_env, dict):
+        conda_env = {}
+    if 'dependencies' not in conda_env:
+        conda_env['dependencies'] = []
+    # Replace python dependency to match minor version
+    python_version = f'{version_info.major}.{version_info.minor}.*'
+    conda_env['dependencies'] = ([f'python={python_version}'] +
+        [spec for spec in  conda_env['dependencies'] if not match_dependency(spec, "python")])
```
Comment (terryyylim) [inline sdk/turing/ensembler.py:389]: super nit: Extra space at `spec in  conda_env ...`

File: sdk/turing/ensembler.py [hunk 7]
```diff
@@ -341,9 +358,33 @@ def create(
...
-            artifact_uri=mlflow.get_artifact_uri()
+            artifact_uri=mlflow.get_artifact_uri(),
+            python_version=python_version
         )
         mlflow.end_run()
 
         return PyFuncEnsembler.from_open_api(
             turing.active_session.create_ensembler(ensembler.to_open_api()))
+
+def _process_conda_env(conda_env):
```
Comment (terryyylim) [inline sdk/turing/ensembler.py:369]: Missing typings here.

**Review by terryyylim — APPROVED** [review]
LGTM. Thank you @krithika369, just a few nitpicks! 🚀

**Pull request merged**